pkg/urbit/jets/e/scow.c

```diff
@@ -0,0 +1,385 @@
+  if (ma != '0')
+    out = u3nc(ma, out);
+  out = u3nc('.', out);
+
+  // suffix the year with a '-' for BC dates
+  if (age == c3n) {
+    out = u3nc('-', out);
+  }
+
+  // The year part is the only place where we have to explicitly loop over the
+  // input because it can be arbitrarily large or small.
+  out = _add_year(year, out);
+
+  out = u3nc('~', out);
+
+  u3z(yod);
+  return out;
+}
+
+static
+u3_noun
+_print_p(u3_atom cor, u3_atom p)
+{
+  // Scramble the raw number to the concealed version.
+  u3_noun hok = u3j_cook("u3we_scow_fein_p", u3k(cor), "fein");
+  u3_atom sxz = u3n_slam_on(hok, u3k(p));
+
+  // Simple galaxy case
+  if (c3y == u3qa_lte(u3k(sxz), 256)) {
+    c3_y a, b, c;
+    po_to_suffix(sxz, &a, &b, &c);
+    u3z(sxz);
+    return u3nq('~', a, b, u3nc(c, 0));
+  }
+
+  u3_atom dyy = u3qc_met(4, u3k(sxz));
+  if (!_(u3a_is_cat(dyy))) {
+    return u3_none;
+  }
+
+  u3_noun list = 0;
+  for (u3_atom imp = 0; imp != dyy; ++imp) {
+    u3_noun log = u3qc_end(4, 1, u3k(sxz));
+    u3_noun prefix = u3qc_rsh(3, 1, u3k(log));
+    u3_noun suffix = u3qc_end(3, 1, log);
+
+    c3_y a, b, c, d, e, f;
+    po_to_prefix(prefix, &a, &b, &c);
+    po_to_suffix(suffix, &d, &e, &f);
+
+    if (imp % 4 == 0) {
+      if (imp != 0) {
+        list = u3nt('-', '-', list);
+      }
+    } else {
+      list = u3nc('-', list);
+    }
+
+    list = u3nq(a, b, c, u3nq(d, e, f, list));
+
+    sxz = u3qc_rsh(4, 1, sxz);
+  }
+
+  u3z(sxz);
+  return u3nc('~', list);
+}
+
+static
+u3_noun
+_print_ud(u3_atom ud)
+{
+  // number of characters printed "between" periods.
+  int between = 0;
+  u3_atom list = 0;
+
+  u3k(ud);
+
+  do {
+    if (between == 3) {
+      list = u3nc('.', list);
+      between = 0;
+    }
+
+    list = u3nc(u3ka_add(u3qa_mod(ud, 10), '0'), list);
+    between++;
+    ud = u3ka_div(ud, 10);
+  } while (c3n == u3r_sing(ud, 0));
+
+  return list;
+}
+
+static
+u3_noun
+_print_uv(u3_atom uv)
+{
+  // number of characters printed "between" periods.
+  int between = 0;
+  u3_atom list = 0;
+
+  u3k(uv);
+
+  do {
+    if (between == 5) {
+      list = u3nc('.', list);
+      between = 0;
+    }
+
+    u3_atom tig = u3qa_mod(uv, 32);
+    list = u3nc(to_digit(tig), list);
+    between++;
+    uv = u3ka_div(uv, 32);
+  } while (c3n == u3r_sing(uv, 0));
+
+  return u3nt('0', 'v', list);
+}
+
+static
+u3_noun
+_print_uw(u3_atom uw)
+{
+  // number of characters printed "between" periods.
+  int between = 0;
+  u3_atom list = 0;
+
+  u3k(uw);
+
+  do {
+    if (between == 5) {
+      list = u3nc('.', list);
+      between = 0;
+    }
+
+    u3_atom tig = u3qa_mod(uw, 64);
+    list = u3nc(to_w_digit(tig), list);
+    between++;
+    uw = u3ka_div(uw, 64);
+  } while (c3n == u3r_sing(uw, 0));
+
+  return u3nt('0', 'w', list);
+}
+
+u3_noun
+u3we_scow(u3_noun cor)
+{
+  u3_noun mod;
+  u3_noun atom;
+
+  if (c3n == u3r_mean(cor, u3x_sam_2, &mod,
+                      u3x_sam_3, &atom, 0) ||
+      !_(u3a_is_cat(mod))) {
```

Comment:
This direct atom check is what I was referring to previously; it doesn't seem like a failure condition to me. IMO, this code should `return u3_none` if the aura is indirect.

Reply:
Also, failing the sample check here is a deterministic error.

Reply:
Given that I've misunderstood what you've requested a few times now, is this what you want?